src/coreclr/jit/lower.cpp

```diff
@@ -7004,6 +7022,205 @@ void Lowering::ContainCheckJTrue(GenTreeOp* node)
+    }
+
+    return count;
+}
+
+//------------------------------------------------------------------------
+// ContainCheckCompareChain : Determine if a chain of ANDs and CMPs can be contained.
+//
+// Arguments:
+//    child - pointer to the node being checked.
+//    parent - parent node of the child.
+//    startOfChain - If found, returns the earliest valid op in the chain.
+//
+bool Lowering::ContainCheckCompareChain(GenTree* child, GenTree* parent, GenTree** startOfChain)
+{
+    assert(parent->OperIs(GT_AND) || parent->OperIs(GT_SELECT));
+
+    if (child->isContainedAndNotIntOrIImmed())
+    {
+        // Already have a chain.
+        *startOfChain = nullptr;
+        return true;
+    }
+    // Can the child be contained.
+    else if (IsSafeToContainMem(parent, child))
+    {
+        if (child->OperIs(GT_AND))
+        {
+            // If Op2 is not contained, then try to contain it.
+            if (!child->AsOp()->gtGetOp2()->isContainedAndNotIntOrIImmed())
+            {
+                if (!ContainCheckCompareChain(child->gtGetOp2(), child, startOfChain))
+                {
+                    // Op2 must be contained in order to contain Op1 or the AND.
+                    return false;
+                }
+            }
+
+            // If Op1 is not contained, then try to contain it.
+            if (!child->AsOp()->gtGetOp1()->isContainedAndNotIntOrIImmed())
+            {
+                if (!ContainCheckCompareChain(child->gtGetOp1(), child, startOfChain))
+                {
+                    return false;
+                }
+            }
+
+            // Contain the AND.
+            child->SetContained();
+            return true;
+        }
+#ifdef TARGET_ARM64
+        else if (child->OperIsCompare())
+        {
+            child->AsOp()->SetContained();
+
+            // Ensure the children of the compare are contained correctly.
+            child->AsOp()->gtGetOp1()->ClearContained();
+            child->AsOp()->gtGetOp2()->ClearContained();
+            ContainCheckConditionalCompare(child->AsOp());
+            *startOfChain = child;
+            return true;
+        }
+#endif
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------
+// ContainCheckAndCompareChain : Determine if an AND is a containable chain
+//
+// Arguments:
+//    node - pointer to the node
+//
+void Lowering::ContainCheckAndChain(GenTree* tree)
+{
+    assert(tree->OperIs(GT_AND));
+
+    if (!comp->opts.OptimizationEnabled())
+    {
+        return;
+    }
+
+    // Determine whether a chain exists.
+    bool inChain    = false;
+    int  chainSize  = 0;
+    int  chainSize2 = CompareChainSize(tree->AsOp()->gtGetOp2(), tree, &inChain);
+    if (chainSize2 > 0 && !inChain)
+    {
+        int chainSize1 = CompareChainSize(tree->AsOp()->gtGetOp1(), tree, &inChain);
+        if (chainSize1 > 0 && !inChain)
+        {
+            chainSize = chainSize1 + chainSize2;
+        }
+    }
```

Comment:
If you're happy with not limiting the maximum size, then it can be simplified. We still need to check there is a valid chain before doing the optimisation.

Reply:
What about changing `ContainCheckCompareChain` to return the number of containments it has marked instead of a bool? That way we can unify `CompareChainSize` and `ContainCheckCompareChain`. We would also need to change `ContainCheckCompareChain` in the cases that the operands are already contained (e.g. skip the `IsSafeToContainMem` there but keep recursing).

Reply:
Problem with that is, for example, an `AND` node, `op2` gives a valid chain, but `op1` is invalid due to being, say, an `ADD` node. That works, but it doesn't give great code:
* op1 generates into a register
* op1 result is moved into flags (there is code in genCodeForBinary to handle this)
* op2 is generated

It seemed better to only generate the chain if `op1` and `op2` of the `AND` are valid. But to that we need a `isValid()` function then a `doIt()` function.

Reply:
I see. And for `GT_SELECT` creating the chain in that case is still going to be better?

Reply:
> I see. And for `GT_SELECT` creating the chain in that case is still going to be better?

For that we probably want to limit the size of the chain (llvm limits this to three).

Reply:
I don't see why it should be limited. If we see `x = (a > 1) & (b > 2) & (c > 3) ... ? n : m`, isn't it always beneficial to use conditional compares, regardless of the length?

Whether transforming short-circuiting and into bitwise and is worth it is another question, that requires PGO information to know.

Reply:
```
cmp a,1 
ble LABEL
cmp  b,2
ble LABEL
cmp c,3
ble LABEL
mov n, m
LABEL:
```

vs:

```
cmp a,1 
ccmp  b,2, le
cmp c,3, le
csel n, m, n, le
```


If a is always less than 1, then it's quicker to run the first example. That's due to the branch predictor.

When we have a single condition (cmp+csel), then using csel gets its best performance as the branch taken or not probability approaches 50/50. (If there was no branch prediction at all then csel would be faster).

With each condition added on to the chain moves the performance closer back down to using branches.

I have a pull request for the performance suite: https://github.com/dotnet/performance/pull/2517. That should probably go in before the select work goes in.

Reply:
Right, that's what I meant with needing PGO information to know. But this kind of transformation (transforming between short-circuiting AND and bitwise AND) would not be happening in the backend. It would have happened earlier.

Here the choice would be between emitting `ccmp` and `and`. I think given that choice, `ccmp` is always worth it regardless of the length of the chain.